x-pack/libbeat/persistentcache/store.go

```diff
@@ -0,0 +1,141 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package persistentcache
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	badger "github.com/dgraph-io/badger/v2"
+
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+// Store is a store for a persistent cache. It can be shared between consumers.
+type Store struct {
+	logger *logp.Logger
+	name   string
+
+	gcQuit chan struct{}
+	db     *badger.DB
+}
+
+// newStore opens a store persisted on the specified directory, with the specified name.
+// If succeeds, returned store must be closed.
+func newStore(logger *logp.Logger, dir, name string) (*Store, error) {
+	dbPath := filepath.Join(dir, name)
+	err := os.MkdirAll(dbPath, 0750)
+	if err != nil {
+		return nil, fmt.Errorf("creating directory for cache store: %w", err)
+	}
+
+	// Opinionated options for the use of badger as a store for metadata caches in Beats.
+	options := badger.DefaultOptions(dbPath)
+	options.Logger = badgerLogger{logger.Named("badger")}
+	// TODO: Disabling sync writes gives better performance, and data loss wouldn't
+	// be a problem for caches. But we are not properly closing processors yet, so let
+	// sync on writes by now.
+	// options.SyncWrites = false
```

Comment:
++

Reply:
Disabling sync on writes in https://github.com/elastic/beats/pull/21754. It shouldn't be needed after #16349.